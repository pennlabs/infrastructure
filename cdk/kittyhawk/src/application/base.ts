import { Construct } from "constructs";
import { removeSubdomain } from "..";
import { Certificate } from "../certificate";
import { Deployment, DeploymentProps } from "../deployment";
import { Ingress, IngressProps } from "../ingress";
import { Service } from "../service";
import { ServiceAccount } from "../serviceaccount";

export interface ApplicationProps {
  /**
   * Ingress configuration
   */
  readonly ingress?: IngressProps;

  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || "undefined_release";
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name,
      });
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      port: props.port,
      ...(props.createServiceAccount
        ? { serviceAccountName: release_name }
        : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      const alreadyCreatedCertificates = new Set<string>();

      props.ingress.rules.forEach((rules) => {
        const finalDomain: string = removeSubdomain(
          rules.host,
          rules.isSubdomain ?? false
        );
        if (!alreadyCreatedCertificates.has(finalDomain)) {
          new Certificate(this, fullname, rules);
          alreadyCreatedCertificates.add(finalDomain);
        }
      });
    }
  }
}
