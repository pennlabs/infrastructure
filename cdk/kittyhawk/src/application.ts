import { Construct } from 'constructs';
import { Deployment, DeploymentProps } from './deployment';
import { Ingress, IngressProps } from './ingress';
import { Service, ServiceProps } from './service'
import { Certificate } from './certificate';

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have 
 * property names that conflict with each other. Typescript may not throw an error and it 
 * could cause problems.
 */
export interface ApplicationProps extends IngressProps, DeploymentProps,
  ServiceProps { }


export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`

    new Service(this, fullname, props);

    new Deployment(this, fullname, props);

    if (props.ingress) {
      new Ingress(this, fullname, props)

      new Certificate(this, fullname, props)
    }
  }
}

/**
 * Helper function that first checks to make sure that the environment variable array 
 * doesn't already contain the env var, then inserts it into the array.
 * @param envArray array of environment variables to insert into 
 * @param envKey name of the environment variable to insert
 * @param envValue value of the environment variable
 */
function insertIfNotPresent(envArray: { name: string, value: string }[], envKey: string, envValue: any) {
  const envSettingsModule = envArray?.filter(env => (env.name === envKey));
  if (envSettingsModule?.length > 0) {
    throw new Error(`${envKey} should not be redefined as an enviroment variable.`)
  }
  envArray.push({ name: envKey, value: envValue });

}
export interface DjangoApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;
}

export class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Have to be careful here with references when mutating things
    let djangoExtraEnv = Array.from(props.extraEnv || []);

    // Insert DJANGO_SETTINGS_MODULE and DOMAIN
    insertIfNotPresent(djangoExtraEnv, 'DJANGO_SETTINGS_MODULE', props.djangoSettingsModule)
    insertIfNotPresent(djangoExtraEnv, 'DOMAIN', props.domain)

    // Configure the ingress using ingressPaths.
    const djangoIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: djangoExtraEnv,
      ingress: djangoIngress,
    });
  }
}

export interface ReactApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {

    // Have to be careful here with references when mutating things
    let reactExtraEnv = Array.from(props.extraEnv || []);

    // Insert DOMAIN and PORT as env vars.
    insertIfNotPresent(reactExtraEnv, 'DOMAIN', props.domain)
    insertIfNotPresent(reactExtraEnv, 'PORT', props.portEnv || '80')

    // Configure the ingress using ingressPaths.
    const reactIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: reactExtraEnv,
      ingress: reactIngress,
    });
  }
}

export class RedisApplication extends Application {
  constructor(scope: Construct, appname: string, redisProps: Partial<ApplicationProps>) {
    super(scope, appname, {
      ...redisProps,
      image: 'redis',
      tag: redisProps.tag || '6.0',
      port: redisProps.port || 6379,
    });
  }
}
