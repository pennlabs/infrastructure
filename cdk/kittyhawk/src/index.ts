
export * from './application';
export * from './autoscaler';
export * from './certificate';
export * from './chart';
export * from './container';
export * from './cronjob';
export * from './deployment';
export * from './ingress';
export * from './service';
