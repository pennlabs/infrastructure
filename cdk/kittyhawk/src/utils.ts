export type NonEmptyArray<T> = [T, ...T[]];

export function nonEmptyMap<T, U>(
  arr: NonEmptyArray<T>,
  callbackfn: (value: T) => U
): NonEmptyArray<U> {
  return arr.map(callbackfn) as any as NonEmptyArray<U>;
}
