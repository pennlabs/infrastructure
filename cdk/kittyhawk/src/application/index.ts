import { Construct } from 'constructs';
import { Certificate } from '../certificate';
import { Deployment, DeploymentProps } from '../deployment';
import { Ingress, IngressProps } from '../ingress';
import { Service } from '../service';
import { ServiceAccount } from '../serviceaccount';
import { nonEmptyMap } from '../utils';

import DjangoApplication from './django';
import ReactApplication from './react';
import RedisApplication from './redis';

export { DjangoApplication };
export { ReactApplication };
export { RedisApplication };

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have
 * property names that conflict with each other. Typescript may not throw an error and it
 * could cause problems.
 */
export interface ApplicationProps {
  readonly ingress?: IngressProps;
  readonly deployment: DeploymentProps;
  readonly port?: number;
  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name
      })
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      ...(props.createServiceAccount ? { serviceAccountName: release_name } : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      nonEmptyMap(props.ingress.rules, rule => {
        return new Certificate(this, fullname, rule);
      });
    }
  }
}