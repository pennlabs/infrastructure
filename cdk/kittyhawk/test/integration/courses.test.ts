import { Construct } from 'constructs';
import cronTime from 'cron-time-generator';
import { CronJob } from '../../src';
import { Application, ReactApplication, RedisApplication } from '../../src/application'
import { chartTest } from '../utils'

export function buildCoursesChart(scope: Construct) {

  /** Penn Courses 
   * https://github.com/pennlabs/penn-courses/blob/master/k8s/values.yaml
  */

  const common = {
    image: 'pennlabs/penn-courses-backend',
    secret: 'penn-courses',
  }

  new RedisApplication(scope, 'redis', { tag: '4.0' })

  new Application(scope, 'celery', {
    ...common,
    cmd: ['celery', 'worker', '-A', 'PennCourses', '-Q', 'alerts,celery', '-linfo'],
    extraEnv: [{ name: 'DJANGO_SETTINGS_MODULE', value: 'PennCourses.settings.production' }],
  })

  new Application(scope, 'backend', {
    ...common,
    cmd: ['celery', 'worker', '-A', 'PennCourses', '-Q', 'alerts,celery', '-linfo'],
    replicas: 3,
    extraEnv: [{ name: 'PORT', value: '80' },
      { name: 'DJANGO_SETTINGS_MODULE', value: 'PennCourses.settings.production' }],
    ingress: [
      { host: 'penncourseplan.com', paths: ['/api', '/admin', '/accounts', '/assets'] },
      { host: 'penncoursealert.com', paths: ['/api', '/admin', '/accounts', '/assets', '/webhook'] },
      { host: 'review.penncourses.org', paths: ['/api', '/admin', '/accounts', '/assets'] },
    ],
  })

  new ReactApplication(scope, 'plan', {
    image: 'pennlabs/pcp-frontend',
    domain: 'penncourseplan.com',
    ingressPaths: ['/'],
  })

  new ReactApplication(scope, 'alert', {
    image: 'pennlabs/pcp-frontend',
    domain: 'penncoursealert.com',
    ingressPaths: ['/'],
  })

  new ReactApplication(scope, 'review', {
    image: 'pennlabs/pcp-frontend',
    domain: 'review.penncourses.org',
    ingressPaths: ['/'],
  })

  new CronJob(scope, 'load-courses', {
    schedule: cronTime.everyDayAt(3),
    ...common,
    cmd: ['python', 'manage.py', 'registrarimport'],
  })

  new CronJob(scope, 'report-stats', {
    schedule: cronTime.everyDayAt(20),
    ...common,
    cmd: ['python', 'manage.py', 'alertstats', '1', '--slack'],
  })
}

test('Penn Courses', () => chartTest(buildCoursesChart));
