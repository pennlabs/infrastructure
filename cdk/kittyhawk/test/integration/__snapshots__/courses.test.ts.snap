// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Penn Courses 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-redis",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-redis",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-redis",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-redis",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-redis",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-redis",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "redis:4.0",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-celery",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-celery",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-celery",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-celery",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-celery",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-celery",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "celery",
                "worker",
                "-A",
                "PennCourses",
                "-Q",
                "alerts,celery",
                "-linfo",
              ],
              "env": Array [
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "PennCourses.settings.production",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "penn-courses",
                  },
                },
              ],
              "image": "pennlabs/penn-courses-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-backend",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-backend",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-backend",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-backend",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 3,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-backend",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-backend",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "celery",
                "worker",
                "-A",
                "PennCourses",
                "-Q",
                "alerts,celery",
                "-linfo",
              ],
              "env": Array [
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "PennCourses.settings.production",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "penn-courses",
                  },
                },
              ],
              "image": "pennlabs/penn-courses-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-backend",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "penncourseplan.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/accounts",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
            ],
          },
        },
        Object {
          "host": "penncoursealert.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/accounts",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/webhook",
              },
            ],
          },
        },
        Object {
          "host": "review.penncourses.org",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/accounts",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "penncourseplan.com",
          ],
          "secretName": "penncourseplan-com-tls",
        },
        Object {
          "hosts": Array [
            "penncoursealert.com",
          ],
          "secretName": "penncoursealert-com-tls",
        },
        Object {
          "hosts": Array [
            "review.penncourses.org",
          ],
          "secretName": "penncourses-org-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncourseplan-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncourseplan.com",
        "*.penncourseplan.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncourseplan-com-tls",
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncoursealert-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncoursealert.com",
        "*.penncoursealert.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncoursealert-com-tls",
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncourses-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncourses.org",
        "*.penncourses.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncourses-org-tls",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-plan",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-plan",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-plan",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-plan",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-plan",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-plan",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DOMAIN",
                  "value": "penncourseplan.com",
                },
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "pennlabs/pcp-frontend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-plan",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "penncourseplan.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-plan",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "penncourseplan.com",
          ],
          "secretName": "penncourseplan-com-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncourseplan-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncourseplan.com",
        "*.penncourseplan.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncourseplan-com-tls",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-alert",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-alert",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-alert",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-alert",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-alert",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-alert",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DOMAIN",
                  "value": "penncoursealert.com",
                },
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "pennlabs/pcp-frontend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-alert",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "penncoursealert.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-alert",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "penncoursealert.com",
          ],
          "secretName": "penncoursealert-com-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncoursealert-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncoursealert.com",
        "*.penncoursealert.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncoursealert-com-tls",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-review",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-review",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-review",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-review",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-review",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-review",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DOMAIN",
                  "value": "review.penncourses.org",
                },
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "pennlabs/pcp-frontend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-review",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "review.penncourses.org",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-review",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "review.penncourses.org",
          ],
          "secretName": "penncourses-org-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncourses-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncourses.org",
        "*.penncourses.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncourses-org-tls",
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-load-courses",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-load-courses",
      "namespace": "default",
    },
    "spec": Object {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "spec": Object {
              "containers": Array [
                Object {
                  "command": Array [
                    "python",
                    "manage.py",
                    "registrarimport",
                  ],
                  "env": Array [
                    Object {
                      "name": "GIT_SHA",
                      "value": "TAG_FROM_CI",
                    },
                  ],
                  "envFrom": Array [
                    Object {
                      "secretRef": Object {
                        "name": "penn-courses",
                      },
                    },
                  ],
                  "image": "pennlabs/penn-courses-backend:TAG_FROM_CI",
                  "imagePullPolicy": "IfNotPresent",
                  "name": "worker",
                  "ports": Array [
                    Object {
                      "containerPort": 80,
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
            },
          },
        },
      },
      "schedule": "0 3 * * *",
      "successfulJobsHistoryLimit": 1,
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-report-stats",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-report-stats",
      "namespace": "default",
    },
    "spec": Object {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "spec": Object {
              "containers": Array [
                Object {
                  "command": Array [
                    "python",
                    "manage.py",
                    "alertstats",
                    "1",
                    "--slack",
                  ],
                  "env": Array [
                    Object {
                      "name": "GIT_SHA",
                      "value": "TAG_FROM_CI",
                    },
                  ],
                  "envFrom": Array [
                    Object {
                      "secretRef": Object {
                        "name": "penn-courses",
                      },
                    },
                  ],
                  "image": "pennlabs/penn-courses-backend:TAG_FROM_CI",
                  "imagePullPolicy": "IfNotPresent",
                  "name": "worker",
                  "ports": Array [
                    Object {
                      "containerPort": 80,
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
            },
          },
        },
      },
      "schedule": "0 20 * * *",
      "successfulJobsHistoryLimit": 1,
    },
  },
]
`;
